### PR TITLE
Added optimize kwarg to torch.einsum

### LINKED ISSUE #58380: Clarity of error message in einsum regressed in performance improvements
## 🐛 Bug

From [Sohrab Samimi on the forums](https://discuss.pytorch.org/t/problem-with-einsum/121476):

```python
first = torch.rand(12,8192,2)
weights1 = torch.rand(12,8192,2)
torch.einsum('bix,iox->box',first,weights1)
```
gives

```
einsum() operands do not broadcast with remapped shapes [original->remapped]: [12, 8192, 2]->[12, 1, 2, 8192] [12, 8192, 2]->[1, 8192, 2, 12]
```

To which they say:
> Can someone explain to me what I am doing wrong?

Clearly, the `i`-dimensions don't match, but the error message sounds relatively opaque to me.

## Expected behavior

In my opinion, it would be better to print the letter/dimension and term number where the size mismatch happens. I think the old version did this, at least sometimes...

@heitorschueroff

### PR DESCRIPTION
**Pull request by heitorschueroff**
Previous PR https://github.com/pytorch/pytorch/pull/56501

This PR adds the `optimize` keyword argument to `torch.einsum` which accepts a contraction path in the same format given by `opt_einsum.contract_path()`. In the future, this keyword argument may be extended to accept more input types to be compatible with `np.einsum`.

fixes https://github.com/pytorch/pytorch/issues/58380

## Benchmark Results

The following results were generated with the random generator from the script below. Note that optimize kwarg should only affect the performance for 3+ operands.

```python 
generate_random_test_cases('cpu', torch.float, seed=2021, n=20, min_ops=3, max_ops=6, min_dims=2, max_dims=3, enable_diagonals=False)
```
```
[----------------------------------------------  ---------------------------------------------]
                                      |  torch.einsum  |  torch.einsum opt  |  numpy.einsum opt
1 threads: ------------------------------------------------------------------------------------
      pnl,Il,nt,GpI,Gpl,Gn            |      10000     |        170         |         280      
      EYX,KGE,EGK,Eo,...Ko            |    6654000     |        120         |         160      
      Vjc,jK,cKV->...                 |         46     |         45         |          81      
      bRX,RyE,EyR,yXE,...Eb,ERb       |        200     |        140         |         230      
      sml,ls,q...j,jql,qm,sjl->...    |       3000     |         73         |         130      
      nI,Or,nl,Trl->nl...I            |         35     |         41         |          58      
      UX,XLF,FH,HL,FX->F...LX         |         50     |         61         |         120      
      YwA,rwC,Cjr,wYr                 |      40000     |        110         |         530      
      lqe,e...V,lq,uql->...           |         48     |         54         |         180      
      BdT,uY...,Yu,IBd,YBI->...       |       3000     |         69         |         100      
      aNn,na,rcn,...Nr                |         59     |         61         |         680      
      iY,YUi,CYi                      |         42     |         53         |          89      
      hQj,wj,UQs                      |        170     |        180         |        1600      
      Is,V...s,spW,ps->pI...          |         67     |         81         |         210      
      fjE,XU,UXf,Ujw,wfE,fX...->f...  |       1500     |         73         |         150      
      WkM,rW,ko,kM...->rM...          |         62     |         69         |         400      
      Dzl,Vb,Kbz,z...b->...           |         66     |         67         |         540      
      TMi,hM,iT,ihM                   |       1200     |         59         |          64      
      pS,iS,oiS,...ib,ibA             |      20000     |        520         |         710      
      rM,MT,hmr,Tm,lmr,r...M->...     |       1500     |        110         |         250      

Times are in microseconds (us).
```

```python
generate_random_test_cases('cuda', torch.float, seed=2021, n=10, n_labels=8, min_ops=4, max_ops=6, min_out_dim=4, max_out_dim=4, min_dims=2, max_size=16, enable_diagonals=False, broadcasting_prob=0)
```
```
[------------------------------------  ------------------------------------]
                                        |  torch.einsum  |  torch.einsum opt
1 threads: -----------------------------------------------------------------
      nGip,Iltp,Gnp,p...l,itI->...nIpG  |      1290      |        256       
      nZS,...ZfS,Se,iq,fe...->Sn...ei   |       205      |        240       
      veMU,eMUC,vCn,wn                  |       206      |        243       
      AvYn,GnYE,nG...A,Afev             |       710      |        246       
      sBb,FbBP,Bs,sF,sB                 |       170      |        200       
      FHV,PTV,P...T,TFI,tVF             |       223      |        257       
      tBK,tf,ur,Bfau->u...KfB           |       150      |        165       
      Bt,tQk,VCt,Vzkt->Qz...tB          |       180      |        171       
      ...LMZ,ZX,XGUL,H...ML             |      5598      |        239       
      Gu...R,Eou,DE,BdGD,RDB,uG         |     58590      |        314       

Times are in microseconds (us).
```

```python 
generate_random_test_cases('cpu', torch.float, seed=2021, n=10, max_ops=2)
```
```
[-----------------------------------  -----------------------------------]
                 |  torch.einsum  |  torch.einsum opt  |  numpy.einsum opt
1 threads: ---------------------------------------------------------------
      tIt,p...t  |        40      |          48        |         220      
      O,aat      |        15      |          20        |          34      
      jWX,qtNt   |       180      |         190        |        3600      
      CC         |         8      |          11        |          20      
      z->...     |         6      |           9        |          23      
      yTMj->...  |         8      |          10        |          30      
      Tvk,JWZ    |     57000      |       57000        |       62000      
      SFsd,s     |       200      |         190        |         110      
      Yx->x...   |         7      |          10        |          24      
      ...y,Ayjr  |      8600      |        8600        |        2200      

Times are in microseconds (us).
```

<details><summary>Benchmark script</summary><p>

```python
# -*- coding: utf-8 -*-
import random
import itertools

import numpy as np
import opt_einsum as oe

import torch
import torch.utils.benchmark as benchmark
from torch.testing._internal.common_utils import make_tensor


class TestCase:
    def __init__(self, operands, sublists, out_sublist=None):
        self.operands = operands
        self.sublists = sublists
        self.out_sublist = out_sublist

    def sublist_format(self, explicit_output=True):
        args = [*itertools.chain(*zip(self.operands, self.sublists))]
        if explicit_output and self.out_sublist:
            args += self.out_sublist
        return args

    def equation_format(self, explicit_output=True):
        equation = ','.join(self._convert_sublist(sublist) for sublist in self.sublists)
        if explicit_output and self.out_sublist:
            equation += '->' + self._convert_sublist(self.out_sublist)
        return (equation, *self.operands)

    def _convert_label(self, label):
        if label == Ellipsis:
            return '...'
        elif label < 26:
            return chr(ord('a') + label)
        else:
            return chr(ord('A') + label - 26)

    def _convert_sublist(self, sublist):
        return ''.join(self._convert_label(label) for label in sublist)


def generate_random_test_cases(
        device, dtype,
        seed=None,
        n=10,
        n_labels=6,
        min_ops=1, max_ops=5,
        min_dims=1, max_dims=4,
        min_size=8, max_size=32,
        min_out_dim=1, max_out_dim=3,
        enable_diagonals=True,
        ellipsis_prob=0.25,
        broadcasting_prob=0.1):

    MAX_LABELS = 52

    assert 0 <= n
    assert 0 <= n_labels < MAX_LABELS
    assert 0 < min_ops <= max_ops
    assert 0 <= min_dims <= max_dims
    assert 0 <= min_size <= max_size
    assert 0 <= max_out_dim
    assert enable_diagonals or max_dims <= n_labels

    if seed:
        random.seed(seed)
        np.random.seed(seed)

    for _ in range(n):

        POSSIBLE_LABELS = np.random.choice(range(MAX_LABELS), n_labels, replace=False)
        LABELS_SIZE = np.random.randint(min_size, max_size + 1, MAX_LABELS)
        ELLIPSIS_SHAPE = np.random.randint(min_size, max_size + 1, max_dims - min_dims)

        operands = []
        sublists = []

        ell_size = 0
        valid_labels = set()

        for _ in range(random.randint(min_ops, max_ops)):
            n_dim = random.randint(min_dims, max_dims)
            labels = np.random.choice(POSSIBLE_LABELS, n_dim, replace=enable_diagonals)
            valid_labels.update(labels)
            shape = LABELS_SIZE[labels]

            mask = np.random.binomial(1, broadcasting_prob, n_dim)
            broadcast_labels = np.unique(labels[mask == 1])
            shape[np.isin(labels, broadcast_labels)] = 1

            labels = list(labels)
            shape = list(shape)

            if n_dim < max_dims and np.random.random() < ellipsis_prob:
                ell_num_dim = random.randint(1, max_dims - n_dim)
                ell_size = max(ell_size, ell_num_dim)
                ell_shape = ELLIPSIS_SHAPE[-ell_num_dim:]
                mask = np.random.binomial(1, broadcasting_prob, ell_num_dim)
                ell_shape[mask == 1] = 1
                ell_index = random.randint(0, n_dim)
                shape[ell_index:ell_index] = ell_shape
                labels.insert(ell_index, Ellipsis)

            operands.append(make_tensor(shape, device, dtype))
            sublists.append(labels)

        out_sublist = None
        if np.random.rand() < 0.5:
            num_out_labels = max(min_out_dim, random.randint(0, min(max_out_dim, len(valid_labels))) - ell_size)
            out_sublist = list(np.random.choice(list(valid_labels), num_out_labels, replace=False))
            out_sublist.insert(random.randint(0, num_out_labels), Ellipsis)

        yield TestCase(operands, sublists, out_sublist).equation_format()


def generate_test_cases_from_issue_57121(device, dtype):
    A = make_tensor((160,) * 2, device, dtype)
    B = make_tensor((160,) * 4, device, dtype)

    yield ('ij,ijkl->kl', A, B)
    yield ('kl,ijkl->ij', A, B)
    yield ('jk,ijkl->il', A, B)
    yield ('il,ijkl->jk', A, B)


def generate_test_cases_from_issue_32591(device, dtype):
    A = make_tensor((1, 1, 16, 2, 16, 2, 16, 2, 2, 2, 2), device, dtype)
    B = make_tensor((729, 1, 1, 2, 1, 2, 1, 2, 2, 2, 2), device, dtype)

    yield ('...ijk,...ijk->...', A, B)

def generate_test_case_from_opt_einsum_example(device, dtype):
    dim = 10
    I = make_tensor((dim, dim, dim, dim), device, dtype)
    C = make_tensor((dim, dim), device, dtype)
    yield ('pi,qj,ijkl,rk,sl->pqrs', C, C, I, C, C)

test_case_generators = [
    generate_random_test_cases('cuda', torch.float, seed=2021, n=10, n_labels=8, min_ops=4, max_ops=6,
                               min_out_dim=4, max_out_dim=4, min_dims=2, max_size=16, enable_diagonals=False, broadcasting_prob=0)
]

results = []

for test_case_generator in test_case_generators:
    for equation, *operands in test_case_generator:
        print('equation:', equation, '\toperands:', [op.shape for op in operands])

        np_ops = [op.cpu().numpy() for op in operands]
        path = oe.contract_path(equation, *np_ops)[0]

        results.append(benchmark.Timer(
            stmt='torch.einsum(equation, *operands)',
            globals={'equation': equation, 'operands': operands},
            sub_label=equation,
            description='torch.einsum',
        ).blocked_autorange())

        results.append(benchmark.Timer(
            stmt='torch.einsum(equation, *operands, optimize=optimize)',
            globals={'equation': equation, 'operands': operands, 'optimize': path},
            sub_label=equation,
            description='torch.einsum opt',
        ).blocked_autorange())

        results.append(benchmark.Timer(
            stmt='numpy.einsum(equation, *operands, optimize=optimize)',
            setup='import numpy',
            globals={'equation': equation, 'operands': np_ops, 'optimize': ('einsum_path', *path)},
            sub_label=equation,
            description='numpy.einsum opt',
        ).blocked_autorange())

compare = benchmark.Compare(results)
compare.trim_significant_figures()
compare.colorize(rowwise=True)
compare.print()
```

</p></details>

### CODE REVIEW

File: aten/src/ATen/native/Linear.cpp [hunk 1]
```diff
@@ -60,8 +60,8 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
   Tensor left = left_;
   Tensor right = right_;
   for (const auto i : c10::irange(dim)) {
-    auto sl = left.size(i)>1;
-    auto sr = right.size(i)>1;
+    auto sl = left.size(i)!=1;
+    auto sr = right.size(i)!=1;
```
Comment (heitorschueroff) [inline aten/src/ATen/native/Linear.cpp:64]: This adds support for zero size dimensions to sumproduct_pair(...)

File: test/test_jit.py [hunk 2]
```diff
@@ -1746,6 +1746,9 @@ def equation_format(x, y):
         def sublist_format(x, y):
             return torch.einsum(x, [0], y, [1], [0, 1])
 
+        def optimize_kwarg(x, y):
+            return torch.einsum('i,j->ij', (x, y), optimize=[0, 1])
```
Comment (heitorschueroff) [inline test/test_jit.py:1750]: For now, JIT script only supports List[int] which matches the signature in native_functions. To support List[Tuple[int, int]] we need to override the JIT function in C++ as was done for sublist format.

File: test/test_linalg.py [hunk 3]
```diff
@@ -4680,7 +4704,7 @@ def test(n=10,                       # how many tests to generate
                 args.append(out_sublist)
                 self._check_einsum(*args, np_args=(equation, *np_operands))
 
-        test(100)
+        test(1000)
```
Comment (mruberry) [inline test/test_linalg.py:4707]: 1000!? How long does this test take in CI now (on CPU and CUDA)?

File: aten/src/ATen/native/native_functions.yaml [hunk 4]
```diff
@@ -1672,7 +1672,7 @@
   dispatch:
     CompositeExplicitAutograd: vdot_out
 
-- func: einsum(str equation, Tensor[] tensors) -> Tensor
+- func: einsum(str equation, Tensor[] tensors, *, int[]? optimize=None) -> Tensor
```
Comment (mruberry) [inline aten/src/ATen/native/native_functions.yaml:1675]: What's the current FC support for adding a new kwarg with a default value? Will old scripted programs that call einsum work OK? If someone exports the new einsum to an older version of PyTorch, will that work?

File: aten/src/ATen/native/Linear.cpp [hunk 5]
```diff
@@ -146,354 +146,324 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
...
 
-  // Convert labels for input operands into an index in [0, 52) and store
-  // them in op_labels for each operand along with ELLIPSIS if present.
-  std::vector<std::vector<uint8_t>> op_labels(num_ops);
-  bool found_ell = false;
-  std::size_t curr_op = 0;
-  for (auto i = decltype(lhs.length()){0}; i < lhs.length(); ++i) {
-    const unsigned char label = lhs[i];
-    switch (label) {
+  if (optimize.has_value()) {
```
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:156]: Comment here explaining this path and the algorithm

File: aten/src/ATen/native/Linear.cpp [hunk 6]
```diff
@@ -146,354 +146,324 @@ static Tensor sumproduct_pair(const Tensor& left_, const Tensor& right_, IntArra
...
-      if (dim_last_op[j] < i) {
-        operand = operand.squeeze(dim);
-        --dim;
-      } else if (dim_last_op[j] == i) {
-        if (result.size(dim) == 1) {
-          operand = operand.sum(dim);
-          result = result.squeeze(dim);
-          --dim;
-        } else {
+    ops.erase(ops.begin() + j);
```
Comment (mruberry) [inline aten/src/ATen/native/Linear.cpp:434]: Super minor style nit: `cbegin`

**Review by mruberry — APPROVED** [review]
This looks really good; would you just validate with Yanan what the current FC support for a new kwarg with a default arg is (for torchscript and mobile)? We (as an org) should write down what's supported somewhere.

A workaround would be to create a new signature that requires the new kwarg explicitly, of course.

File: test/test_linalg.py (hunk 3 above)
Comment (heitorschueroff) [reply test/test_linalg.py:4707]: Around 2-3s

**Pull request closed without merging**